### ISSUE TITLE
testing v2.0.0: Cloud init problem with piped multiline yaml

### ISSUE THREAD

**Issue by axgkl** [vitobotta/hetzner-k3s#388]
### Discussed in https://github.com/vitobotta/hetzner-k3s/discussions/385

Hi 

this here is supported in yaml:

```yaml
  - ip route add 169.254.0.0/16 via 172.31.1.1
  - mkdir -p /etc/network/interfaces.d
  - |
    IFACE=$(ip -o -4 addr list | grep ' 10.4.' | cut -d ' ' -f 2 | xargs)
    fn='/etc/network/interfaces.d/'$IFACE''
    echo 'auto '$IFACE'' > $fn
    echo 'iface '$IFACE' inet dhcp' >> $fn
    echo '    post-up ip route add default via 10.4.0.1' >> $fn
    echo '    post-up ip route add 169.254.169.254 via 172.31.1.1' >> $fn
 ```
 
very useful for keeping some state over command sequences in cloud init.
 
But when I use such a constract as postcommand, it's butchered into sth like this:

```
 
 
 (...)
packages: ['fail2ban', 'wireguard', 'ifupdown']

runcmd:
- hostnamectl set-hostname $(curl http://169.254.169.254/hetzner/v1/metadata/hostname)
- update-crypto-policies --set DEFAULT:SHA1 || true
(...)
- mkdir -p /etc/network/interfaces.d
- IFACE=$(ip -o -4 addr list | grep ' 10.4.' | cut -d ' ' -f 2 | xargs)
fn='/etc/network/interfaces.d/'$IFACE''
echo 'auto '$IFACE'' > $fn
echo 'iface '$IFACE' inet dhcp' >> $fn
echo '    post-up ip route add default via 10.4.0.1' >> $fn
echo '    post-up ip route add 169.254.169.254 via 172.31.1.1' >> $fn

- ...
```

And fails with schema error.

Note not for you but for readers: One can see the produced cloud init on the server here: 

```
# root@ax-master2:/var/lib/cloud/instances/50866657# cat user-data.txt  
```

No stress, can workaround via writing a file and reading it every time I need that interface name.

Note: Last week I did NOT use such a construct, so I do NOT know if the bug had been present in the old version.

**Comment by axgkl** [comment]
And now I see this: https://github.com/vitobotta/hetzner-k3s/issues/109#issuecomment-1793219812

and am confused. I try that "one big pipe" style.

**Comment by axgkl** [comment]
Nope - second command breaks it:

![2024-07-25_1990x288_scrot](https://github.com/user-attachments/assets/f849fc6c-4df4-4b5c-88ef-c0462bbcdb9c)

**Comment by axgkl** [comment]
Hmm, does also not work - which would mean that it's a problem of v2 if it worked in v1.

Minimum test case:

```yaml
post_create_commands:
- |
  cat > /setup <<EOF
  printf "started" > status
  export foo=bar
  timedatectl set-timezone Europe/Berlin
  printf "done" > status
  touch "$foo"
  EOF
- chmod +x /setup
- /setup
```

(that works when I paste it into the UI)

But the installer creates on the server this

```
(...)
runcmd:
- hostnamectl set-hostname $(curl http://169.254.169.254/hetzner/v1/metadata/hostname)
- update-crypto-policies --set DEFAULT:SHA1 || true
- echo "nameserver 8.8.8.8" > /etc/k8s-resolv.conf
- cat > /setup <<EOF
printf "started" > status
export foo=bar
timedatectl set-timezone Europe/Berlin
printf "done" > status
touch "$foo"
EOF

- chmod +x /setup
- /setuproot@ax-master1:/var/lib/cloud/instances/50869948#
```

which does fail also when pasted like that, into the UI:

![2024-07-25_882x492_scrot](https://github.com/user-attachments/assets/3d8aadf1-68c9-4267-b8dc-b0284a6e2cd4)

---

I see a lot of code around those post_create_commands in the source. imho, if such constructs are hard to parse, maybe allowing to create files (which one could simply chmod and execute in runcmd would be a simple solution).

**Comment by axgkl** [comment]
Workaround (for others with that problem): Keep the state in a file, e.g. my iface name:

```
  - ip -o -4 addr list | grep ' 10.4.' | cut -d ' ' -f 2 | xargs > /iface
  - IFACE=$(cat /iface) ; fn='/etc/network/interfaces.d/'$IFACE''
  - IFACE=$(cat /iface) ; echo 'auto '$IFACE'' > $fn
  - IFACE=$(cat /iface) ; echo 'iface '$IFACE' inet dhcp' >> $fn
```

**Comment by vitobotta** [comment]
Are you sure you are using the right number of spaces? Or are you using tabs in your editor?

I do like the idea of `post_create_scripts` as it would make things easier. I am taking a note about it but I prefer doing that in 2.0.1 as I want ton finalize 2.0.0 soon.

**Comment by vitobotta** [comment]
This should be fixed in rc2. Can you double check when the binary for your OS is ready? https://github.com/vitobotta/hetzner-k3s/actions/runs/10098594405

**Comment by axgkl** [comment]
It worked, also the EOF style.

---

Impressive build pipeline btw.

This would be cool btw, to adapt the the new versions in each build:

```
root@bastion:~# ./hetzner-k3s --version
1.1.5
```

**Comment by vitobotta** [comment]
BTW one note: in one snippet you had `$foo`. Remember to escape the $ sign otherwise it will be used for interpolation in the yaml itself when creating the file with the multiline command :)
